### PR TITLE
Add virtual merge job

### PR DESCRIPTION
**Pull request by jcoyne**
Ref #1458

### CODE REVIEW

File: spec/services/virtual_merge_tool_spec.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+            <resource sequence="2" id="#{namespaceless}_2" type="image">
+              <relationship type="alsoAvailableAs" objectId="#{child2.id}"/>
+            </resource>
+          </contentMetadata>
+        XML
+        child1.reload
+        expect(child1.object_relations[:is_constituent_of]).to eq ["info:fedora/#{parent.id}"]
+        child2.reload
+        expect(child2.object_relations[:is_constituent_of]).to eq ["info:fedora/#{parent.id}"]
+      end
```
Comment (ndushay) [inline spec/services/virtual_merge_tool_spec.rb]: I think we might want tests for:
- parent isn't open for modification
- parent ok, but one child isn't ... show other child added?
- useful error information going into logs for diff errors (to ultimately be viewed by user)

File: app/services/virtual_merge_tool.rb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    @parent ||= begin
+      Dor::Item.find(@parent_druid).tap do |item|
+        raise "Parent item #{item.pid} is not open for modification" unless item.allows_modification?
+      end
+    end
+  end
+
+  # merge all the child resources into the parent as virtual links
+  def merge(child)
+    raise "Child item #{child.pid} is not open for modification" unless child.allows_modification?
```
[the remark this answers is not in the document]
Comment (jcoyne) [reply app/services/virtual_merge_tool.rb]: This is the existing behavior. Should we change it?

[the remark this answers is not in the document]
Comment (ndushay) [reply app/services/virtual_merge_tool.rb]: ah - good point.  I dunno - @blalbrit @andrewjbtw  -- wanna chime in?  Do we want to fail the JOB if a child isn't open for modification, or do we want to fail just that one child?

[the remark this answers is not in the document]
Comment (andrewjbtw) [reply app/services/virtual_merge_tool.rb]: Going by the motivating use case for virtual merge - combining individual maps that make up an atlas - it sounds like failing one child and moving on could result in the equivalent of an atlas that's missing a page. So I think failing the job makes sense if one child object can't be added to the parent, especially if there's a risk of publishing an incomplete object.

**Comment by jcoyne** [comment]
@ndushay I've changed the base branch for this PR to be the "virtual-merge" branch.

File: app/services/constituent_service.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+# TODO: this should move to dor-services-app and dor-services-client
```
Comment (jcoyne) [inline app/services/constituent_service.rb]: Moving to dor-services-app: https://github.com/sul-dlss/dor-services-app/pull/321

File: bin/virtual-merge [hunk 4]
```diff
@@ -102,6 +65,7 @@ logger.level = options[:debug].nil? ? Logger::INFO : Logger::DEBUG
 
 # run the tool
 require_relative '../config/environment'
-VirtualMergeTool.new(logger, ARGV.delete_at(0),
-                     options[:input].nil? ? ARGV : File.readlines(options[:input]).map(&:strip)).run(options[:purge])
+VirtualMergeTool.new(logger: logger,
```
[the remark this answers is not in the document]
Comment (ndushay) [reply bin/virtual-merge]: Per conversation with @andrewjbtw this morning, we may be running the script as a way of dealing with the backload of spreadsheets that have been piling up for 2 years.  So we need to keep the script.  Hopefully, once you re-add it to this PR, you'll do some manual testing to be sure the refactoring doesn't break it (I like that the new way allows for tests, and affords the use of the same code for the script and for the bulk action.

[the remark this answers is not in the document]
Comment (jcoyne) [reply bin/virtual-merge]: Okay, I've just left it as is.

**Comment by jcoyne** [comment]
@ndushay I've just restored `bin/virtual-merge` and dropped a comment in about it being deprecated.  Yes, this should be blocked until we get https://github.com/sul-dlss/dor-services-client/pull/76 into a release and we pin that release in the argo Gemfile.

**Pull request merged**